lib/sendserver.c

```diff
@@ -43,6 +43,72 @@ static int rc_check_reply(AUTH_HDR *, int, char const *, unsigned char const *,
  * @{
  */
 
+/** Encodes CHAP password and stores it in an output buffer 
+ *
+ * @param vp a pointer to a VALUE_PAIR.
+ * @param auth a pointer to AUTH_HDR.
+ * @param id an identifer to use for encryption.
+ * @param output an output buffer to store encrypted chap password string
+ * @return value 0 on success
+ */
+static int rc_chap_encode(VALUE_PAIR * vp, AUTH_HDR * auth, int id, uint8_t * output)
+{
+    /* variables */
+    uint8_t    *ptr;
+    uint8_t    str_val[(AUTH_STRING_LEN + 1) * 2 + 1];
+    int        str_len;
+    VALUE_PAIR *password;
+    VALUE_PAIR *challenge;
+    VALUE_PAIR *vp_ptr = vp;
+
+    /* sanity check */
+    if (vp_ptr == NULL || output == NULL)
+    {
+            rc_log(LOG_ERR, "rc_chap_encode: unintialized pointer");
+            return -1;
+    }
+
+    str_len = 0;
+    ptr     = str_val;
+
+    /* Add chap identifier */
+    *ptr++  = id;
+    str_len++;
+
+    /* Retrieve unencrypted CHAP-password */
+    vp_ptr   = vp;
+    password = rc_avpair_get(vp_ptr, PW_CHAP_PASSWORD, 0);
+    if (password) {
+            memcpy(ptr, password->strvalue, password->lvalue);
+            ptr     += password->lvalue;
+            str_len += password->lvalue;
+    } else {
+           rc_log(LOG_ERR, "rc_chap_encode: cannot retrieve chap password attribute");
+           return -1;
+    }
+
+    /* Use CHAP-challenge pair, otherwise use request-auhenticator */
+    vp_ptr    = vp;
+    challenge = rc_avpair_get(vp_ptr, PW_CHAP_CHALLENGE, 0);
+    if (challenge) {
+            memcpy(ptr, challenge->strvalue, challenge->lvalue);
+            str_len += challenge->lvalue;
+    } else {
+            if(auth) {
```

Comment:
In that case we don't set the PW_CHAP_CHALLENGE. Will that be a problem for the server, or it is not required?

Reply:
According to the RFC it seems that either the CHAP-Challenge is used or if not present then the Request Authenticator is used instead:

https://datatracker.ietf.org/doc/html/rfc2865#section-5.3 :

      The CHAP challenge value is found in the CHAP-Challenge Attribute
      (60) if present in the packet, otherwise in the Request
      Authenticator field.

So seems as standard requires?